File: src/trl/vec.rs
use super::raw_vec::RawVec;
use super::rng::Rng;

pub struct Vec<T> {
    buf: RawVec<T>,
    len: usize,
}

impl<T> Vec<T> {
    pub fn new() -> Self {
        Vec {
            buf: RawVec::new(),
            len: 0,
        }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        let buf = RawVec::with_capacity(capacity);
        Vec { buf, len: 0 }
    }

    pub fn from<const N: usize>(arr: &[T; N]) {
        let len = arr.len();
        let vec = ::std::vec::Vec::from([1, 2, 3]);
        let buf = RawVec::with_capacity(len);
        unsafe {
            buf.from_raw_parts(arr, len);
        }
        Vec { buf, len }
    }

    pub fn with_len(len: usize, default: T) -> Self {
        let buf = RawVec::with_capacity(len);
        Vec { buf, len }
    }

    pub fn capacity(&self) -> usize {
        self.buf.capacity()
    }
}

impl<T> Rng for Vec<T> {
    // fn left(&self) -> Opt<T> {
    //     return self.buf.get(0);
    // }

    // fn right(&self) -> Opt<T> {
    //     return self.arr.get(self.arr.len() - 1);
    // }

    // fn popLeft() -> (Vec<T>, Opt<T>) {
    //     let first = this.arr[0];
    //     return (Vec.new(rest), first);
    // }
}

//     popRight (): [Vec<T>, T | Und] {
//         const last = this._arr[this._arr.length - 1]
//         return [Vec.new(this._arr.slice(1)), last as T | Und]
//     }

//     pop (): [Vec<T>, T | Und] {
//         return this.popRight()
//     }

//     at (key: usize): T | Und {
//         return this._arr[key]
//     }

//     len (): usize {
//         return this._arr.length
//     }

//     every (fn: (item: T, key: usize) => bool): bool {
//         return this._arr.every(fn)
//     }

//     some (fn: (item: T, key: usize) => bool): bool {
//         return this._arr.some(fn)
//     }

//     fold<R> (initialValue: R, fn: (acc: R, item: T, index: usize) => R): R {
//         return this._arr.reduce<R>(fn, initialValue)
//     }

//     reduce (fn: (a: T, b: T, index: usize) => T): T {
//         return this._arr.reduce(fn)
//     }

//     map<R> (fn: (a: T, key: usize) => R): Vec<R> {
//         return Vec.new(this._arr.map(fn))
//     }

//     filter (fn: (a: T, key: usize) => bool): Vec<T> {
//         return Vec.new(this._arr.filter(fn))
//     }

//     find (fn: (a: T, key: usize) => bool): [value: T, key: usize] | Und {
//         const key = this._arr.findIndex(fn)
//         return key === -1 ? und : [this._arr[key], key]
//     }

//     includes (item: T): bool {
//         return this.find((a) => a === item) != und
//     }

//     each<R> (fn: (a: T, key: usize) => R): Und {
//         return this._arr.forEach(fn)
//     }

//     slice (start: usize | Und, end: usize | Und): Vec<T> {
//         return Vec.new(this._arr.slice(start!, end!))
//     }

//     push (val: T): Vec<T> {
//         return Vec.new([...this._arr, val])
//     }

//     join (val: Str): Str {
//         return Str.new(this._arr.join(val._()))
//     }

//     concat (val: Vec<T>): Vec<T> {
//         return Vec.new([...this._arr, ...val._()])
//     }

//     _ (): T[] {
//         return this._arr
//     }

//     valueOf (): T[] {
//         return this._()
//     }

//     toString (): T[] {
//         return this._()
//     }
// }

File: src/vec.ts
import { List } from 'immutable'
import { Opt, optFrom, or } from './opt'
import { RandomAccessFiniteRng } from './rng'

import { Slice } from './slice'
import { Str } from './str'

export class Vec<T> implements RandomAccessFiniteRng<usize, T> {
    private readonly _arr: List<T>

    private constructor(list: List<T>) {
        this._arr = list
    }

    static new<T>(len = 0, init = 0): Vec<T> {
        const arr = Array(len)
        for (let i = 0; i < len; i++) {
            arr[i] = init
        }
        return new Vec(List(arr))
    }

    static from<T>(arr: T[]): Vec<T> {
        return new Vec(List(arr))
    }

    left(): Opt<T> {
        return optFrom(this._arr.get(0))
    }

    right(): Opt<T> {
        return optFrom(this._arr.get(this.len() - 1))
    }

    popLeft(): [Vec<T>, Opt<T>] {
        const first = optFrom(this._arr.first())
        const rest = new Vec(this._arr.shift())
        return [rest, first]
    }

    popRight (): [Vec<T>, Opt<T>] {
        const last = optFrom(this._arr.last())
        const rest = new Vec(this._arr.pop())
        return [rest, last]
    }

    skipLeft (amount: usize): Vec<T> {
        return new Vec(this._arr.skip(amount))
    }

    skipRight (amount: usize): Vec<T> {
        return new Vec(this._arr.skipLast(amount))
    }

    slice (fn: (len: usize) => Slice<usize>): Vec<T> {
        const len = this.len()
        const slice = fn(len)
        return new Vec(this._arr.slice(or(slice.left(), 0), or(slice.right(), len)))
    }

    len (): usize {
        return this._arr.size
    }

    isEmpty(): bool {
        return this.len() === 0
    }

    pushRight (val: T): Vec<T> {
        return new Vec(this._arr.push(val))
    }

    pushLeft (val: T): Vec<T> {
        return new Vec(this._arr.unshift(val))
    }

    get (index: usize): Opt<T> {
        return optFrom(this._arr.get(index))
    }

    set (index: usize, val: T): Vec<T> {
        return new Vec(this._arr.set(index, val))
    }

    join (val: Str): Str {
        return Str.from(this._arr.join(val.inner()))
    }

    toString (): string {
        return this._arr.toArray().toString()
    }
}
